Common/src/main/java/com/blamejared/crafttweaker/api/tag/manager/ITagManager.java

```diff
@@ -1,187 +1,158 @@
-    default boolean exists(String name) {
-        
-        return exists(new ResourceLocation(name));
+    default boolean exists(String id) {
+
+        ResourceLocation rl = ResourceLocation.tryParse(id);
+        Objects.requireNonNull(rl, "Invalid Resource Location passed! Given: '" + id + "'");
+        return exists(rl);
     }
-    
-    /**
-     * Checks if a tag already exists. Does the same as calling `.exists` on a tag directly
-     *
-     * @param location The resource location to check for
-     *
-     * @return Whether this tag already exists
-     */
+
     @ZenCodeType.Method
     @ZenCodeType.Operator(ZenCodeType.OperatorType.CONTAINS)
-    default boolean exists(ResourceLocation location) {
-        
-        return getTagCollection().getAllTags().containsKey(location);
+    default boolean exists(ResourceLocation id) {
+
+        return tagMap().containsKey(id);
     }
-    
-    /**
-     * Retrieves a list of all tags currently registered.
-     */
+
     @ZenCodeType.Method
-    @ZenCodeType.Getter("all")
-    default List<MCTag<T>> getAllTags() {
-        
-        return getTagCollection().getAllTags()
-                .keySet()
-                .stream()
-                .map(tagLocation -> new MCTag<>(tagLocation, this))
-                .collect(Collectors.toList());
+    @ZenCodeType.Operator(ZenCodeType.OperatorType.CONTAINS)
+    default boolean exists(MCTag<T> tag) {
+
+        return exists(tag.id());
     }
-    
-    /**
-     * Retrieves all tags contain the provided element
-     *
-     * @param element The element whose tags should be returned
-     */
+
     @ZenCodeType.Method
-    default List<MCTag<T>> getAllTagsFor(T element) {
-        //While this is a working implementation, using TagCollection#getOwningTags would probably be more efficient ^^
-        //Therefore the CrT TagManagers override this impl
-        return getAllTags().stream()
-                .filter(tag -> tag.contains(element))
-                .collect(Collectors.toList());
+    @ZenCodeType.Getter("tagMap")
+    Map<ResourceLocation, MCTag<T>> tagMap();
+
+    @ZenCodeType.Method
+    @ZenCodeType.Getter("tags")
+    default List<MCTag<T>> tags() {
+
+        return new ArrayList<>(tagMap().values());
     }
-    
-    /**
-     * Get the tag type. In a Bracket call, this will used to determine which ITagManager to use.
-     * <p>
-     * {@code <tag:{tag_type}:tag_location:tag_name>} <br>
-     * {@code <tagManager:{tag_type}>}
-     */
+
     @ZenCodeType.Method
-    @ZenCodeType.Getter("tagType")
-    String getTagFolder();
-    
-    /**
-     * Just so you won't have to implement that yourself ^^
-     */
+    default List<MCTag<T>> getTagsFor(T element) {
+
+        return tags().stream().filter(tag -> tag.contains(element)).toList();
+    }
+
     @Override
     default String getCommandString() {
-        
-        return "<tagmanager:" + getTagFolder() + ">";
+
+        return "<tagmanager:" + tagFolder() + ">";
```

Comment:
Mods with custom registries might break this, we should look into it.

Reply:
We haven't ran into an issue with it so far and vanilla is *hopefully* going to enforce "no custom tag directories" soon